### ISSUE TITLE
Access to price scale

### ISSUE THREAD

**Issue by mattou78400** [tradingview/lightweight-charts#621]
Hi,

I can't find a way to get access to the displayed min/max of the price scale/axis
It'd be useful to get several graphs "in-sync" in terms of axis.

How should I go about this? Or where is this information stored?
All I can find is the margin in %, but it doesn't help me much... 

Thanks a lot!

**Comment by mattou78400** [comment]
Best way I found was to use the max and min of the displayed candles
But if I manually resize the scale, I can't get its extremas...

**Comment by timocov** [comment]
There is no such api yet. There is tricky moment when percentage mode is enabled - you cannot set visible range in that mode, and what units you'd expect to see there in the getter? percentage? prices? Also, for instance, setting visible range should reset `autoScale` option, because they are exclusive.

**Comment by mattou78400** [comment]
Hey, thanks for your answer

I'm expecting to get the max and min PRICE of the price scale
e.g. 
![image](https://user-images.githubusercontent.com/39678117/99383355-09ef8c00-28ce-11eb-934c-13544b9dcb09.png)

It should return me 16900 (ish) and 16570 here

I'd like to be able to slide/scale around the graph view and get those extreme value to display further info, "in sync" with what I'm seeing in the lightweight chart

Thanks

**Comment by timocov** [comment]
> I'd like to be able to slide/scale around the graph view and get those extreme value to display further info, "in sync" with what I'm seeing in the lightweight chart

Yeah, I got your point. You also want to be notified when price range is changed manually I guess...

> It should return me 16900 (ish) and 16570 here

It's easy answer for normal mode. What's about [percentage](https://github.com/tradingview/lightweight-charts/blob/master/docs/price-scale.md#percentage-scales)? Visible values there aren't the same as prices. And setting price range in this mode is forbidden - we need to either throw an error or disable percentage mode in this case.

**Comment by mattou78400** [comment]
I have to be honest, I didn't look into percentage mode much.
But I believe it'd make sense to send back the extremas as well
![image](https://user-images.githubusercontent.com/39678117/99384121-41126d00-28cf-11eb-93fa-a701050da511.png)
taking this example, something like:

{min: -0.6, max: 0.9}

> Yeah, I got your point. You also want to be notified when price range is changed manually I guess...
Absolutely right, that'd be great!

thanks a lot

**Comment by timocov** [comment]
> {min: -0.6, max: 0.9}

Getting that value, what's that? Percentage values? Price values? Normalized percentage values? I just want to say that it's a tricky question and we cannot solve only your particular case and don't solve others.

> Intuitively, why not adding such info to "visibileRange"?

Because visibleRange is visible **time** range, it even exists in TimeScale API and returns `TimeRange` not `VisibleRange` or so.

**Comment by mattou78400** [comment]
> Getting that value, what's that? Percentage values? Price values? Normalized percentage values? I just want to say that it's a tricky question and we cannot solve only your particular case and don't solve others.

I understand, it's simply the max and min displayed on the axis (percentage values, as displayed) so I can know the range displayed on the graph (in % for % or in terms of price)

**Comment by timocov** [comment]
> so I can know the range displayed on the graph (in % for % or in terms of price)

Yes, this is your particular use case :) I can imagine a situation when a user wants to sync 2 charts, one with percentage more -> another without percentage mode, and is this case we need to have some converter from percentage values to price ones... Anyway, as I said, this is tricky and we need to wait for some other feedback before implement this API.

**Comment by mattou78400** [comment]
for the percentage mode, we can also send back the prices extremas as those should be 

max_price = (price at displayed origine) * (1 + max_percentage)
min_price = (price at displayed origine) * (1 + min_percentage)

with price at displayed origine, being the actual price at 0%

Hope this helps.

Meanwhile, are you aware of any "hack" that could help me get those price extremas?

Thanks!

**Comment by mattou78400** [comment]
I was told a solution for my use case
Simply make use of coordinateToPrice with coordinates 0 and "chart height"

Thanks

**Comment by hanhoungsoo** [comment]
<div class="tradingview-widget-container">
  <div id="tradingview_a9e65"></div>
  <div class="tradingview-widget-copyright">TradingView 제공 <a href="https://kr.tradingview.com/symbols/NASDAQ-AAPL/" rel="noopener" target="_blank"><span class="blue-text">AAPL 차트</span></a></div>
  <script type="text/javascript" src="https://s3.tradingview.com/tv.js"></script>
  <script type="text/javascript">
  new TradingView.widget(
  {
  "width": 980,
  "height": 610,
  "symbol": "NASDAQ:AAPL",
  "interval": "D",
  "timezone": "Etc/UTC",
  "theme": "light",
  "style": "1",
  "locale": "kr",
  "toolbar_bg": "#f1f3f6",
  "enable_publishing": false,
  "allow_symbol_change": true,
  "container_id": "tradingview_a9e65"
}
  );
  </script>
</div>

**Comment by quixot1c** [comment]
A possible solution would be to provide both the absolute and relative values in the event. Something like
```
{
  price: 3.14,
  percentage: 0.5
}
```
Then the consumer of the event can decide whether they need the absolute or relative value and drive the view of another graph that way.

**Comment by timocov** [comment]
> Then the consumer of the event can decide whether they need the absolute or relative value and drive the view of another graph that way.

@quixot1c how to handle this when you try to change price range?

**Comment by quixot1c** [comment]
@timocov when you change the price range on a particular graph, ie by zooming or scrolling, you compute which percentage of the total range is in view and report both the absolute price range and percentage range. So maybe not just one value like I said before but more like:
```
{
  event: "range changed",
  price: [2.50, 5.00],
  percentage: [0.1, 0.8]
}
```
Now I haven't looked at your API in depth so I really don't know how you format events and such. Just wanted to quickly pitch in when I read this issue and thought of a high level solution. I can't comment on the exact implementation.

**Comment by thanhlmm** [comment]
My thought on this is: Do we have any internal function to convert from price range to percentage and vice versa? By implement a conversate function, we would have only 1 source of trust then everything should be easy on what ever scale type.

I might not know any issue if we do that. How do you guys think?

**Comment by timocov** [comment]
@thanhlmm to convert to percentage scale additionally you need to have `firstValue` (the very first value across all series attached to the scale). Also, it is impossible for now to set price range when percentage mode is enabled because of the fact that percentage mode enables auto mode and you cannot disable it. The reason on this behaviour is UX mostly, because enabling manual mode with percentage scale leads mess on the chart when you move it.

So we can allow to set price range (in values), in the way by disabling autoscale automatically and enabling "normal" mode from percentage/indexed to 100 modes, if that fine for everybody. And the same for getters - the library will return price range (in values), by we will provide a method to convert prices to percentages if you need them.

But! In case of "percentage" (or indexed to 100 - they are almost the same) mode, the thing is that all series attached to the price scale will use percentage values to place and draw their values (convert values to coordinate) rather than their own "values" (prices) (this is basically the nature of these modes). I don't know how tricky it might be for you guys, but I admit that it might be.

**Comment by CodeSmart101** [comment]
Has there been any update on this issue ?
Is it possible to get the visible min and max price from the price scale ? Even if it means to manually make calls to the lightweight charts to get the price, I am OK with it.
Without this functionality, anything related to overlays becomes improbable.

**Comment by 0x33dm** [comment]
> Has there been any update on this issue ? Is it possible to get the visible min and max price from the price scale ? Even if it means to manually make calls to the lightweight charts to get the price, I am OK with it. Without this functionality, anything related to overlays becomes improbable.

indeed this is very needed, I'm also having issues with scaling the price

**Comment by MustafaZayed** [comment]
Just in case someone wants to do this. There is a workaround that worked pretty well with me. The idea is to create a dummy plugin to provide this functionality for you:

create some `.ts` file and the below code. this is just some basic code (modify if needed):
```
import { CanvasRenderingTarget2D } from 'fancy-canvas';
import {
  ISeriesPrimitive,
  ISeriesPrimitiveAxisView,
  ISeriesPrimitivePaneRenderer,
  ISeriesPrimitivePaneView,
  SeriesPrimitivePaneViewZOrder,
  Time,
} from 'lightweight-charts';

class PaneRenderer implements ISeriesPrimitivePaneRenderer {
  constructor(private onUpdate: () => void) {}

  draw(target: CanvasRenderingTarget2D): void {
    this.onUpdate();
  }
  drawBackground?(target: CanvasRenderingTarget2D): void {
    this.onUpdate();
  }
}

class PaneView implements ISeriesPrimitivePaneView {
  constructor(private onUpdate: () => void) {}
  zOrder?(): SeriesPrimitivePaneViewZOrder {
    return 'top';
  }
  renderer(): ISeriesPrimitivePaneRenderer | null {
    return new PaneRenderer(this.onUpdate);
  }
}

class TimeAxisView implements ISeriesPrimitiveAxisView {
  constructor(private onUpdate: () => void) {}
  coordinate(): number {
    return 0;
  }
  fixedCoordinate?(): number | undefined {
    return 0;
  }
  text(): string {
    return '';
  }
  textColor(): string {
    return '';
  }
  backColor(): string {
    return '';
  }
  visible?(): boolean {
    return true;
  }
  tickVisible?(): boolean {
    return true;
  }
  update() {
    this.onUpdate();
  }
}

export class CustomPrimitive implements ISeriesPrimitive<Time> {
  _paneViews: ISeriesPrimitivePaneView[];
  _timeAxisViews: ISeriesPrimitiveAxisView[];
  constructor(private updateCallback: () => void) {
    this._paneViews = [new PaneView(updateCallback)];
    this._timeAxisViews = [new TimeAxisView(updateCallback)];
  }

  updateAllViews() {
    this.updateCallback();
  }
  timeAxisViews() {
    return this._timeAxisViews;
  }
  paneViews() {
    return this._paneViews;
  }
  priceAxisPaneViews(): readonly ISeriesPrimitivePaneView[] {
    return this._paneViews;
  }
}
```

Then you can use it like this:
```
const primitive = new CustomPrimitive(() => {
      // some call back function that will execute if
      // the price scale change or time scale change
});
candlestickSeries.attachPrimitive(primitive);
```

If you want it to execute only in case of price scale change, then Modify the `CustomPrimitive` class to be (did not test it but should work):

```
export class CustomPrimitive implements ISeriesPrimitive<Time> {
  _paneViews: ISeriesPrimitivePaneView[];
  constructor(private updateCallback: () => void) {
    this._paneViews = [new PaneView(updateCallback)];
  }

  updateAllViews() {
    this.updateCallback();
  }
  paneViews() {
    return this._paneViews;
  }
  priceAxisPaneViews(): readonly ISeriesPrimitivePaneView[] {
    return this._paneViews;
  }
}
```

**Comment by garmoshka-mo** [comment]
Here is how to set price range:
```
  chart.setPriceRange = (minValue, maxValue) => {
    chart.mySeries.applyOptions({
      autoscaleInfoProvider: () => ({
        priceRange: {minValue, maxValue},
      })
    })
  }
```
